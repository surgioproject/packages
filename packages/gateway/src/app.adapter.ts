import { ExpressAdapter } from '@nestjs/platform-express'
import express from 'express'

export function createAdapter(): ExpressAdapter {
  const instance = express()
  return new ExpressAdapter(instance)
}
