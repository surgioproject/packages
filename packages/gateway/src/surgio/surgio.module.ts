import { Module, DynamicModule, Global } from '@nestjs/common'
import path from 'path'

import { loadConfig } from 'surgio/build/utils/config'
import { KEY, SurgioHelper } from './surgio-helper'
import { SurgioService } from './surgio.service'

@Global()
@Module({})
export class SurgioModule {
  public static register(options: { readonly cwd: string }): DynamicModule {
    return {
      module: SurgioModule,
      providers: [
        SurgioService,
        {
          provide: KEY,
          useFactory: (): Promise<SurgioHelper> => {
            const configFile = path.join(options.cwd, 'surgio.conf.js')
            const config = loadConfig(options.cwd, configFile)
            const helper = new SurgioHelper(options.cwd, config)

            return helper.init()
          },
        },
      ],
      exports: [SurgioService],
    }
  }
}
