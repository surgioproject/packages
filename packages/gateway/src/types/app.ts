import { Request } from 'express'

import { Role } from '../constants/role'

export interface UserContext {
  readonly accessToken: string
  readonly roles: Array<Role>
}

export interface EnrichedRequest extends Request {
  user?: UserContext
}
