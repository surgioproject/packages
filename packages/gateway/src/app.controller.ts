import {
  Controller,
  Get,
  Res,
  Param,
  Query,
  HttpException,
  HttpStatus,
  UseGuards,
  Req,
  Logger,
} from '@nestjs/common'
import { Request, Response } from 'express'
import { Artifact } from 'surgio/generator'
import _ from 'lodash'
import { getUrl } from 'surgio/utils'
import { URL } from 'url'
import dayjs from 'dayjs'
import duration from 'dayjs/plugin/duration'
import NodeCache from 'node-cache'

import { APIAuthGuard } from './auth/api-auth.guard'
import { Roles } from './auth/roles.decorator'
import { Role } from './constants/role'
import { SurgioService } from './surgio/surgio.service'

dayjs.extend(duration)

const resCache = new NodeCache({
  maxKeys: 100,
  stdTTL: dayjs.duration({ days: 7 }).asSeconds(),
})

@Controller()
@UseGuards(APIAuthGuard)
export class AppController {
  private readonly logger = new Logger(AppController.name)

  constructor(private readonly surgioService: SurgioService) {}

  @Get('/get-artifact/:name')
  @Roles(Role.VIEWER)
  public async getArtifact(
    @Res() res: Response,
    @Param() params: { readonly name: string },
    @Query() query: GetArtifactQuery,
    @Req() req: Request
  ): Promise<void> {
    const dl = query.dl
    const format = query.format
    const filter = query.filter
    const urlParams = _.omit(query, ['dl', 'format', 'filter', 'access_token'])
    const artifactName: string = params.name
    const userAgent = req.headers['user-agent']
    let artifact: string | undefined | Artifact
    let isCache = false

    try {
      artifact =
        format !== void 0
          ? await this.surgioService.transformArtifact(
              artifactName,
              format,
              filter
            )
          : await this.surgioService.getArtifact(artifactName, {
              downloadUrl: new URL(
                req.url,
                this.surgioService.config.publicUrl
              ).toString(),
            })
    } catch (err) {
      if (resCache.has(req.url)) {
        isCache = true
        artifact = resCache.get(req.url) as string

        this.logger.warn('Artifact 生成错误，使用缓存')
        this.logger.warn(err.stack || err)
      } else {
        throw err
      }
    }

    if (typeof artifact !== 'undefined') {
      res.header('content-type', 'text/plain; charset=utf-8')
      res.header('cache-control', 'private, no-cache, no-stores')

      if (dl === '1') {
        res.header(
          'content-disposition',
          `attachment; filename="${artifactName}"`
        )
      }

      this.logger.warn(
        `[download-artifact] ${artifactName} "${userAgent || '-'}"`
      )

      await this.sendPayload(
        req,
        res,
        artifact,
        {
          ...urlParams,
          userAgent: userAgent || '',
        },
        isCache
      )
    } else {
      throw new HttpException('NOT FOUND', HttpStatus.NOT_FOUND)
    }
  }

  @Get('/export-providers')
  @Roles(Role.VIEWER)
  public async exportProvider(
    @Req() req: Request,
    @Res() res: Response,
    @Query() query: ExportProviderQuery
  ): Promise<void> {
    const providers: string[] = query.providers
      ? query.providers.split(',').map((item) => item.trim())
      : []
    const userAgent = req.headers['user-agent']

    if (!providers.length) {
      throw new HttpException(
        '参数 provider 必须指定至少一个值',
        HttpStatus.BAD_REQUEST
      )
    }

    providers.forEach((provider) => {
      if (!this.surgioService.surgioHelper.providerMap.has(provider)) {
        throw new HttpException(
          `provider ${provider} 不存在`,
          HttpStatus.NOT_FOUND
        )
      }
    })

    const format = query.format
    const template = query.template

    if (!format && !template) {
      throw new HttpException(
        '参数 format 和 template 必须指定至少一个值',
        HttpStatus.BAD_REQUEST
      )
    }

    const dl = query.dl
    const filter = query.filter
    const urlParams = _.omit(query, [
      'dl',
      'format',
      'template',
      'filter',
      'access_token',
      'providers',
    ])
    let artifact: Artifact | string
    let isCache = false

    try {
      if (format) {
        artifact = await this.surgioService.exportProvider(
          providers[0],
          format,
          undefined,
          {
            filter,
            ...(providers.length > 1
              ? {
                  combineProviders: providers.splice(1),
                }
              : null),
          }
        )
      } else {
        artifact = await this.surgioService.exportProvider(
          providers[0],
          undefined,
          template,
          {
            filter,
            downloadUrl: new URL(
              req.url as string,
              this.surgioService.config.publicUrl
            ).toString(),
            ...(providers.length > 1
              ? {
                  combineProviders: providers.splice(1),
                }
              : null),
          }
        )
      }
    } catch (err) {
      if (resCache.has(req.url)) {
        isCache = true
        artifact = resCache.get(req.url) as string

        this.logger.warn('Provider 导出错误，使用缓存')
        this.logger.warn(err.stack || err)
      } else {
        throw err
      }
    }

    if (artifact instanceof Artifact) {
      res.header('content-type', 'text/plain; charset=utf-8')
      res.header('cache-control', 'private, no-cache, no-stores')

      if (dl === '1') {
        res.header(
          'content-disposition',
          `attachment; filename="${artifact.artifact.name}"`
        )
      }

      this.logger.warn(
        `[download-artifact] ${artifact.artifact.name} "${userAgent || '-'}"`
      )

      await this.sendPayload(
        req,
        res,
        artifact,
        {
          ...urlParams,
          userAgent: userAgent || '',
        },
        isCache
      )
    } else {
      await this.sendPayload(req, res, artifact, undefined, isCache)
    }
  }

  @Get('/render')
  @Roles(Role.VIEWER)
  public async renderTemplate(
    @Req() req: Request,
    @Res() res: Response,
    @Query() query: RenderTemplateQuery
  ): Promise<void> {
    const { template } = query
    const config = this.surgioService.config
    const gatewayConfig = config?.gateway
    const gatewayHasToken = !!gatewayConfig?.accessToken

    if (!template) {
      throw new HttpException(
        '参数 template 必须指定一个值',
        HttpStatus.BAD_REQUEST
      )
    }

    try {
      res.header('content-type', 'text/plain; charset=utf-8')
      res.header(
        'cache-control',
        `s-maxage=${60 * 60 * 24}, stale-while-revalidate`
      )

      const html = this.surgioService.surgioHelper.templateEngine.render(
        `${template}.tpl`,
        {
          downloadUrl: new URL(
            req.url as string,
            this.surgioService.config.publicUrl
          ).toString(),
          getUrl: (p: string) =>
            getUrl(
              config.publicUrl,
              p,
              gatewayHasToken ? gatewayConfig?.accessToken : undefined
            ),
        }
      )

      res.send(html)
    } catch (err) {
      if (err.message.includes('template not found')) {
        throw new HttpException('NOT FOUND', HttpStatus.NOT_FOUND)
      } else {
        throw err
      }
    }
  }

  private async sendPayload(
    req: Request,
    res: Response,
    artifact: string | Artifact,
    urlParams?: Record<string, string>,
    isCachedPayload?: boolean
  ): Promise<void> {
    const config = this.surgioService.config
    const gatewayConfig = config?.gateway

    if (typeof artifact === 'string') {
      if (gatewayConfig?.useCacheOnError && !isCachedPayload) {
        resCache.set(req.url, artifact)
      }

      if (isCachedPayload) {
        res.header('x-use-cache', 'true')
      }

      res.send(artifact)
    } else {
      // 只支持输出单个 Provider 的流量信息
      if (!isCachedPayload && artifact.providerMap.size === 1) {
        const providers = artifact.providerMap.values()
        const provider = providers.next().value

        if (provider.supportGetSubscriptionUserInfo) {
          try {
            const subscriptionUserInfo =
              await provider.getSubscriptionUserInfo()

            if (subscriptionUserInfo) {
              const values = ['upload', 'download', 'total', 'expire'].map(
                (key) => `${key}=${subscriptionUserInfo[key] || 0}`
              )

              res.header('subscription-userinfo', values.join('; '))
            }
          } catch (err) {
            this.logger.error('处理订阅信息失败')
            this.logger.error(err.stack || err)
          }
        }
      }

      const body = artifact.render(undefined, {
        ...(urlParams ? this.processUrlParams(urlParams) : undefined),
      })

      if (gatewayConfig?.useCacheOnError && !isCachedPayload) {
        resCache.set(req.url, body)
      }

      if (isCachedPayload) {
        res.header('x-use-cache', 'true')
      }

      res.send(body)
    }
  }

  private processUrlParams(
    urlParams: Record<string, string>
  ): Record<string, string> {
    const result: NonNullable<any> = Object.create(null)

    Object.keys(urlParams).forEach((key) => {
      _.set(result, key, urlParams[key])
    })

    return result
  }
}

interface GetArtifactQuery {
  readonly format: string
  readonly filter?: string
  readonly dl?: string
  readonly access_token?: string
  readonly [key: string]: string | undefined
}

interface RenderTemplateQuery {
  readonly template: string
}

interface ExportProviderQuery {
  readonly providers: string
  readonly format?: string
  readonly template?: string
  readonly filter?: string
  readonly dl?: string
  readonly access_token?: string
  readonly [key: string]: string | undefined
}
