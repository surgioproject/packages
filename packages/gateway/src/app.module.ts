import { MiddlewareConsumer, Module, RequestMethod } from '@nestjs/common'
import { ConfigModule, ConfigService } from '@nestjs/config'
import { join } from 'path'
import { ServeStaticModule } from '@nestjs/serve-static'
import getPackage from '@surgio/gateway-frontend'

import { AppController } from './app.controller'
import { ApiModule } from './api/api.module'
import { CookieParserMiddleware } from './middleware/cookie-parser.middleware'
import { PrepareMiddleware } from './middleware/prepare.middleware'
import { SurgioModule } from './surgio/surgio.module'
import { SurgioService } from './surgio/surgio.service'
import { AuthModule } from './auth/auth.module'
import configuration from './config/configuration'

const FE_MODULE = require.resolve('@surgio/gateway-frontend')
const frontendPackage = getPackage()

@Module({
  imports: [
    ServeStaticModule.forRoot({
      rootPath: join(FE_MODULE, '../build'),
      serveStaticOptions: {
        cacheControl: true,
        etag: true,
        maxAge: '31d',
        setHeaders: (res, path) => {
          if (
            path.endsWith('.js') ||
            path.endsWith('.css') ||
            path.endsWith('.json')
          ) {
            res.setHeader('x-frontend-version', frontendPackage.version)
          }
        },
      },
    }),
    ConfigModule.forRoot({
      ignoreEnvFile: true,
      isGlobal: true,
      load: [configuration],
    }),
    SurgioModule.register({
      cwd: process.env.SURGIO_PROJECT_DIR || process.cwd(),
    }),
    ApiModule,
    AuthModule,
  ],
  controllers: [AppController],
})
export class AppModule {
  constructor(private surgioService: SurgioService) {}

  configure(consumer: MiddlewareConsumer): void {
    const secret = this.surgioService.surgioHelper.configHash

    CookieParserMiddleware.configure(secret)
    consumer.apply(CookieParserMiddleware).forRoutes('*')

    consumer
      .apply(PrepareMiddleware)
      .exclude({ path: 'render', method: RequestMethod.ALL })
      .forRoutes(AppController)
  }
}
