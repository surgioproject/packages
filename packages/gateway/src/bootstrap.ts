import { NestFactory } from '@nestjs/core'
import { NestExpressApplication } from '@nestjs/platform-express'
import express from 'express'

import { AppModule } from './app.module'
import { AppExceptionsFilter } from './filter/app-exception.filter'
import { createAdapter } from './app.adapter'

export async function bootstrap(): Promise<NestExpressApplication> {
  const app = await NestFactory.create<NestExpressApplication>(
    AppModule,
    createAdapter(),
    {
      logger:
        process.env.NODE_ENV === 'production' ? ['error', 'warn'] : ['verbose'],
    }
  )

  applyMiddlwares(app)

  return app
}

export function applyMiddlwares(
  app: NestExpressApplication
): NestExpressApplication {
  app.useGlobalFilters(new AppExceptionsFilter())
  app.use(express.json())

  return app
}
