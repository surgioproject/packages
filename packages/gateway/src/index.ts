import 'source-map-support/register'
import { ConfigService } from '@nestjs/config'
import { createServer, IncomingMessage, OutgoingMessage, Server } from 'http'
import serverless, { Handler } from 'serverless-http'

import { bootstrap } from './bootstrap'

export const createHttpServer = (): Server => {
  const prepare = bootstrap().then((nestApp) => {
    return nestApp.init()
  })

  return createServer((req: IncomingMessage, res: OutgoingMessage) => {
    prepare.then((nestApp) => {
      nestApp.getHttpServer().emit('request', req, res)
    })
  })
}

export const startServer = (): Promise<Server> => {
  let port: number

  return bootstrap()
    .then((app) => {
      const configService = app.get<ConfigService>(ConfigService)
      port = configService.get('port') as number

      return app.listen(port)
    })
    .then((httpServer: Server) => {
      console.log('> Your app is ready at http://127.0.0.1:' + port)
      return httpServer
    })
    .catch((err) => {
      console.error('⚠️ 服务启动失败')
      console.error(err)
      process.exit(1)
    })
}

export const createLambdaHandler = () => {
  let handler: Handler

  return async (event, context) => {
    if (!handler) {
      const nestApp = await bootstrap()
      await nestApp.init()
      const adapter = nestApp.getHttpAdapter().getInstance()
      handler = serverless(adapter)
    }

    return await handler(event, context)
  }
}

export const bootstrapServer = bootstrap
