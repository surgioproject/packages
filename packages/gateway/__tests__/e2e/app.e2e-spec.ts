import { NestFactory } from '@nestjs/core'
import { NestExpressApplication } from '@nestjs/platform-express'
import supertest from 'supertest'
import Bluebird from 'bluebird'

import { AppModule } from '../../src/app.module'
import { applyMiddlwares } from '../../src/bootstrap'
import { SurgioService } from '../../src/surgio/surgio.service'

describe('AppController (e2e)', () => {
  let app: NestExpressApplication
  let token
  let viewerToken

  beforeAll(async () => {
    app = await NestFactory.create(AppModule, { logger: false })
    applyMiddlwares(app)

    const surgioService = app.get<SurgioService>(SurgioService)
    token = surgioService.config.gateway?.accessToken
    viewerToken = surgioService.config.gateway?.viewerToken

    await app.init()
  })

  afterAll(async () => {
    await app.close()
  })

  test('/ (GET)', async () => {
    await supertest(app.getHttpServer()).get('/').expect(200)
  })

  test('/get-artifact (GET)', async () => {
    const res = await supertest(app.getHttpServer()).get(
      `/get-artifact/test.conf?access_token=${token}`
    )

    expect(res.status).toBe(200)
    expect(res.get('subscription-userinfo')).toBeUndefined()
    expect(res.text).toMatchSnapshot()
  })

  test('/get-artifact (GET) with viewerToken', async () => {
    const res = await supertest(app.getHttpServer()).get(
      `/get-artifact/test.conf?access_token=${viewerToken}`
    )

    expect(res.status).toBe(200)
    expect(res.get('subscription-userinfo')).toBeUndefined()
    expect(res.text).toMatchSnapshot()
  })

  test('/get-artifact (GET) attachment', async () => {
    await supertest(app.getHttpServer())
      .get(`/get-artifact/test.conf?access_token=${token}&dl=1`)
      .expect('content-disposition', 'attachment; filename="test.conf"')
  })

  test('/get-artifact (GET) userinfo header', async () => {
    const res = await supertest(app.getHttpServer()).get(
      `/get-artifact/test3.conf?access_token=${token}`
    )

    expect(res.get('subscription-userinfo')).toBe(
      'upload=891332010; download=29921186546; total=322122547200; expire=1586330887'
    )
  })

  test('/get-artifact (GET) 404', async () => {
    await supertest(app.getHttpServer())
      .get(`/get-artifact/notfound.conf?access_token=${token}`)
      .expect(404)
  })

  test('/get-artifact (GET) unauthorized supertest', async () => {
    await supertest(app.getHttpServer())
      .get(`/get-artifact/test.conf`)
      .expect(401)
    await supertest(app.getHttpServer())
      .get(`/get-artifact/notfound.conf?access_token=wrong`)
      .expect(401)
  })

  test('/get-artifact (GET) custom params', async () => {
    {
      const res = await supertest(app.getHttpServer()).get(
        `/get-artifact/custom-params.conf?access_token=${token}`
      )

      expect(res.status).toBe(200)
      expect(res.text).toMatchSnapshot()
    }

    {
      const res = await supertest(app.getHttpServer()).get(
        `/get-artifact/custom-params.conf?access_token=${token}&foo=new&child.bar=new`
      )

      expect(res.status).toBe(200)
      expect(res.text).toMatchSnapshot()
    }
  })

  test('/get-artifact (GET) error cache', async () => {
    await supertest(app.getHttpServer())
      .get(`/get-artifact/clash-error-after-first.conf?access_token=${token}`)
      .expect(200)

    await Bluebird.delay(500)

    const res = await supertest(app.getHttpServer()).get(
      `/get-artifact/clash-error-after-first.conf?access_token=${token}`
    )

    expect(res.status).toBe(200)
    expect(res.get('x-use-cache')).toBe('true')
  })

  test('customParams should not contaminate prototype', async () => {
    await supertest(app.getHttpServer()).get(
      `/get-artifact/custom-params.conf?access_token=${token}&constructor.prototype.hacked=bar`
    )
    expect(({} as any).hacked).toBeUndefined()
  })

  test('/export-providers (GET)', async () => {
    const res = await supertest(app.getHttpServer()).get(
      `/export-providers?access_token=${token}&providers=clash&format=surge-policy`
    )

    expect(res.status).toBe(200)
    expect(res.text).toMatchSnapshot()
  })

  test('/export-providers (GET) with viewerToken', async () => {
    const res = await supertest(app.getHttpServer()).get(
      `/export-providers?access_token=${viewerToken}&providers=clash&format=surge-policy`
    )

    expect(res.status).toBe(200)
    expect(res.text).toMatchSnapshot()
  })

  test('/export-providers (GET) multiple providers', async () => {
    const res = await supertest(app.getHttpServer()).get(
      `/export-providers?access_token=${token}&providers=clash,custom&format=surge-policy`
    )

    expect(res.status).toBe(200)
    expect(res.text).toMatchSnapshot()
  })

  test('/export-providers (GET) multiple providers 404', async () => {
    const res = await supertest(app.getHttpServer()).get(
      `/export-providers?access_token=${token}&providers=clash,custom,notfound&format=surge-policy`
    )

    expect(res.status).toBe(404)
    expect(res.text).toMatchSnapshot()
  })

  test('/export-providers (GET) global filter', async () => {
    const res = await supertest(app.getHttpServer()).get(
      `/export-providers?access_token=${token}&providers=clash,custom&format=surge-policy&filter=customFilters.globalFilter`
    )

    expect(res.status).toBe(200)
    expect(res.text).toMatchSnapshot()
  })

  test('/export-providers (GET) internal filter', async () => {
    const res = await supertest(app.getHttpServer()).get(
      `/export-providers?access_token=${token}&providers=clash,custom&format=surge-policy&filter=hkFilter`
    )

    expect(res.status).toBe(200)
    expect(res.text).toMatchSnapshot()
  })

  test('/export-providers (GET) private filter', async () => {
    const res = await supertest(app.getHttpServer()).get(
      `/export-providers?access_token=${token}&providers=clash,custom&format=surge-policy&filter=customFilters.testFilter`
    )

    expect(res.status).toBe(200)
    expect(res.text).toMatchSnapshot()
  })

  test('/export-providers (GET) using template', async () => {
    const res = await supertest(app.getHttpServer()).get(
      `/export-providers?access_token=${token}&providers=clash,custom&template=export`
    )

    expect(res.status).toBe(200)
    expect(res.text).toMatchSnapshot()
  })

  test('/export-providers (GET) using wrong template', async () => {
    await supertest(app.getHttpServer())
      .get(
        `/export-providers?access_token=${token}&providers=clash,custom&template=notfound`
      )
      .expect(500)
  })

  test('/export-providers (GET) unauthorized supertest', async () => {
    await supertest(app.getHttpServer())
      .get(`/export-providers?providers=clash`)
      .expect(401)
    await supertest(app.getHttpServer())
      .get(`/export-providers?access_token=wrong&providers=clash`)
      .expect(401)
  })

  test('/render (GET)', async () => {
    const res = await supertest(app.getHttpServer()).get(
      `/render?access_token=${token}&template=render`
    )

    expect(res.status).toBe(200)
    expect(res.text).toMatchSnapshot()
  })

  test('/render (GET) with viewerToken', async () => {
    const res = await supertest(app.getHttpServer()).get(
      `/render?access_token=${viewerToken}&template=render`
    )

    expect(res.status).toBe(200)
    expect(res.text).toMatchSnapshot()
  })

  test('/render (GET) sub folder', async () => {
    const res = await supertest(app.getHttpServer()).get(
      `/render?access_token=${token}&template=sub-folder%2Frender`
    )

    expect(res.status).toBe(200)
    expect(res.text).toMatchSnapshot()
  })

  test('/render (GET) not found', async () => {
    await supertest(app.getHttpServer())
      .get(`/render?access_token=${token}&template=render-not-found`)
      .expect(404)
  })
})
