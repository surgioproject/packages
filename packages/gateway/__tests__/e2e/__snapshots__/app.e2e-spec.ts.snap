// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppController (e2e) /export-providers (GET) 1`] = `
"ss1 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=true
ss2 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=false, obfs=tls, obfs-host=www.bing.com
vmess = vmess, server, 443, username=uuid, vmess-aead=true
http 1 = https, server, 443, username, password, tls13=false, skip-cert-verify=false
http 2 = http, server, 443, username, password
snell = snell, server, 44046, psk=yourpsk, obfs=http
ss4 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=false, obfs=tls, obfs-host=example.com"
`;

exports[`AppController (e2e) /export-providers (GET) global filter 1`] = `"🇺🇸 US = ss, us.example.com, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=true, obfs=tls, obfs-host=gateway-carry.icloud.com, tfo=true"`;

exports[`AppController (e2e) /export-providers (GET) internal filter 1`] = `""`;

exports[`AppController (e2e) /export-providers (GET) multiple providers 1`] = `
"ss1 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=true
ss2 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=false, obfs=tls, obfs-host=www.bing.com
vmess = vmess, server, 443, username=uuid, vmess-aead=true
http 1 = https, server, 443, username, password, tls13=false, skip-cert-verify=false
http 2 = http, server, 443, username, password
snell = snell, server, 44046, psk=yourpsk, obfs=http
ss4 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=false, obfs=tls, obfs-host=example.com
🇺🇸 US = ss, us.example.com, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=true, obfs=tls, obfs-host=gateway-carry.icloud.com, tfo=true
Snell = snell, us.example.com, 443, psk=password, obfs=tls
HTTPS = https, us.example.com, 443, username, password"
`;

exports[`AppController (e2e) /export-providers (GET) multiple providers 404 1`] = `"{"status":"error","statusCode":404,"error":"provider notfound 不存在"}"`;

exports[`AppController (e2e) /export-providers (GET) private filter 1`] = `
"ss1 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=true
ss2 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=false, obfs=tls, obfs-host=www.bing.com
vmess = vmess, server, 443, username=uuid, vmess-aead=true
ss4 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=false, obfs=tls, obfs-host=example.com
🇺🇸 US = ss, us.example.com, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=true, obfs=tls, obfs-host=gateway-carry.icloud.com, tfo=true"
`;

exports[`AppController (e2e) /export-providers (GET) using template 1`] = `
"https://example.com/export-providers?access_token=abcd&providers=clash,custom&template=export

ss1 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=true
ss2 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=false, obfs=tls, obfs-host=www.bing.com
vmess = vmess, server, 443, username=uuid, vmess-aead=true
http 1 = https, server, 443, username, password, tls13=false, skip-cert-verify=false
http 2 = http, server, 443, username, password
snell = snell, server, 44046, psk=yourpsk, obfs=http
ss4 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=false, obfs=tls, obfs-host=example.com
🇺🇸 US = ss, us.example.com, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=true, obfs=tls, obfs-host=gateway-carry.icloud.com, tfo=true
Snell = snell, us.example.com, 443, psk=password, obfs=tls
HTTPS = https, us.example.com, 443, username, password
"
`;

exports[`AppController (e2e) /export-providers (GET) with viewerToken 1`] = `
"ss1 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=true
ss2 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=false, obfs=tls, obfs-host=www.bing.com
vmess = vmess, server, 443, username=uuid, vmess-aead=true
http 1 = https, server, 443, username, password, tls13=false, skip-cert-verify=false
http 2 = http, server, 443, username, password
snell = snell, server, 44046, psk=yourpsk, obfs=http
ss4 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=false, obfs=tls, obfs-host=example.com"
`;

exports[`AppController (e2e) /get-artifact (GET) 1`] = `
"🇺🇸US 1 = ss, us.example.com, 443, encrypt-method=chacha20-ietf-poly1305, password=password, obfs=tls, obfs-host=gateway-carry.icloud.com
🇺🇸US 2 = ss, us.example.com, 443, encrypt-method=chacha20-ietf-poly1305, password=password
🇺🇸 US = ss, us.example.com, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=true, obfs=tls, obfs-host=gateway-carry.icloud.com, tfo=true
Snell = snell, us.example.com, 443, psk=password, obfs=tls
HTTPS = https, us.example.com, 443, username, password
ss1 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=true
ss2 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=false, obfs=tls, obfs-host=www.bing.com
vmess = vmess, server, 443, username=uuid, vmess-aead=true
http 1 = https, server, 443, username, password, tls13=false, skip-cert-verify=false
http 2 = http, server, 443, username, password
snell = snell, server, 44046, psk=yourpsk, obfs=http
ss4 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=false, obfs=tls, obfs-host=example.com
----
🇺🇸US 1, 🇺🇸US 2, 🇺🇸US 3, 🇺🇸 US, Snell, HTTPS, ss1, ss2, ss3, vmess, http 1, http 2, snell, ss4, ss-wss
"
`;

exports[`AppController (e2e) /get-artifact (GET) custom params 1`] = `
"original
----
original
"
`;

exports[`AppController (e2e) /get-artifact (GET) custom params 2`] = `
"new
----
new
"
`;

exports[`AppController (e2e) /get-artifact (GET) with viewerToken 1`] = `
"🇺🇸US 1 = ss, us.example.com, 443, encrypt-method=chacha20-ietf-poly1305, password=password, obfs=tls, obfs-host=gateway-carry.icloud.com
🇺🇸US 2 = ss, us.example.com, 443, encrypt-method=chacha20-ietf-poly1305, password=password
🇺🇸 US = ss, us.example.com, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=true, obfs=tls, obfs-host=gateway-carry.icloud.com, tfo=true
Snell = snell, us.example.com, 443, psk=password, obfs=tls
HTTPS = https, us.example.com, 443, username, password
ss1 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=true
ss2 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=false, obfs=tls, obfs-host=www.bing.com
vmess = vmess, server, 443, username=uuid, vmess-aead=true
http 1 = https, server, 443, username, password, tls13=false, skip-cert-verify=false
http 2 = http, server, 443, username, password
snell = snell, server, 44046, psk=yourpsk, obfs=http
ss4 = ss, server, 443, encrypt-method=chacha20-ietf-poly1305, password=password, udp-relay=false, obfs=tls, obfs-host=example.com
----
🇺🇸US 1, 🇺🇸US 2, 🇺🇸US 3, 🇺🇸 US, Snell, HTTPS, ss1, ss2, ss3, vmess, http 1, http 2, snell, ss4, ss-wss
"
`;

exports[`AppController (e2e) /render (GET) 1`] = `
"#!name=MitM All Hostnames
#!desc=Perform MitM on all hostnames with port 443, except those to Apple and other common sites which can't be inspected. You still need configure CA certificate and enable the main switch of MitM.
#https://example.com/render?access_token=abcd&template=render

[MITM]
hostname = -*.apple.com, -*.icloud.com, -*.mzstatic.com, -*.crashlytics.com, -*.facebook.com, -*.instagram.com, *
"
`;

exports[`AppController (e2e) /render (GET) sub folder 1`] = `
"#!name=MitM All Hostnames
#!desc=Perform MitM on all hostnames with port 443, except those to Apple and other common sites which can't be inspected. You still need configure CA certificate and enable the main switch of MitM.
#https://example.com/render?access_token=abcd&template=sub-folder%2Frender

[MITM]
hostname = -*.apple.com, -*.icloud.com, -*.mzstatic.com, -*.crashlytics.com, -*.facebook.com, -*.instagram.com, *
"
`;

exports[`AppController (e2e) /render (GET) with viewerToken 1`] = `
"#!name=MitM All Hostnames
#!desc=Perform MitM on all hostnames with port 443, except those to Apple and other common sites which can't be inspected. You still need configure CA certificate and enable the main switch of MitM.
#https://example.com/render?access_token=efgh&template=render

[MITM]
hostname = -*.apple.com, -*.icloud.com, -*.mzstatic.com, -*.crashlytics.com, -*.facebook.com, -*.instagram.com, *
"
`;
